manager/common/license.go

```diff
@@ -0,0 +1,204 @@
+	}
+	log.Printf("License released by %s \n", hash)
+	c.licenses[vendor].Used -= num
+	return c.licenses[vendor].Total
+}
+
+func (c *LicenseCounter) peekLicense(vendor string) (int, int, bool) {
+	c.counterMutex.Lock()
+	defer c.counterMutex.Unlock()
+	return c.licenses[vendor].Used, c.licenses[vendor].Total, c.licenses[vendor].Used < c.licenses[vendor].Total
+}
+
+func (c *LicenseCounter) enqueue(key string, val *Client) {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	c.queue[val.Vendor] = append(c.queue[val.Vendor], key)
+	c.clients[key] = val
+	log.Printf("Client %s from %s@%s waiting for %d %s %s \n",
+		key, val.User, val.IP, val.Quantity, val.Vendor, val.Feature)
+	log.Printf("%s job queue: %s \n", val.Vendor, strings.Join(c.queue[val.Vendor], ", "))
+}
+
+func (c *LicenseCounter) dequeue(key string, vendor string) {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	c.queue[vendor] = c.queue[vendor][1:]
+	delete(c.clients, key)
+	log.Printf("Client %s removed from the %s queue \n", key, vendor)
+}
+
+func (c *LicenseCounter) del(key string, vendor string) {
+	index := -1
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	for i := 0; i < len(c.queue[vendor]); i++ {
+		if c.queue[vendor][i] == key {
+			c.queue[vendor] = append(c.queue[vendor][:i], c.queue[vendor][i+1:]...)
+			delete(c.clients, key)
+			log.Printf("Job cancelled by client %s - removed from queue \n", key)
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		log.Printf("%s is not in queue \n", key)
+	}
+}
+
+func (c *LicenseCounter) validLicense(vendor string, feature string) bool {
+	licenses := map[string]map[string]bool{"xilinx": {"vivado": true},
+                                           "cadence": {"xcelium": true}}
+	_, ok := licenses[vendor][feature]
+	return ok
+}
+
+func (c *LicenseCounter) peekQueue(vendor string, hash string) bool {
+	c.clientMutex.Lock()
+	defer c.clientMutex.Unlock()
+	return c.queue[vendor][0] == hash
+}
+
+type Client struct {
+	IP       string
+	User     string
+	Start    time.Time
+	Quantity int32
+	Vendor   string
+	Feature  string
+}
+
+type License struct {
+	Total int
+	Used  int
+}
+
+var totalCadenceLic = 3
+var totalXilinxLic = 3
+var licenseCounter = LicenseCounter{
+	licenses: map[string]*License{"xilinx": &License{Used: 0, Total: totalXilinxLic}, "cadence": &License{Used: 0, Total: totalCadenceLic}},
+	clients:  map[string]*Client{},
+	queue:    map[string][]string{"xilinx": make([]string, 0), "cadence": make([]string, 0)}}
+
+type Server struct {
+	rng *rand.Rand
+}
+
+func (s *Server) Polling(stream rpc_license.License_PollingServer) error {
+	var hash string
+	var vendor string
+	var status error
+	s.rng = rand.New(rand.NewSource(rand.Int63()))
+	for {
+		recv, err := stream.Recv()
+		// remove client hash from queue when connection is broken
+		if err == io.EOF {
+			licenseCounter.del(hash, vendor)
+			return nil
+		}
+		if err != nil {
+			licenseCounter.del(hash, vendor)
```

Comment:
this seems a bit suspicious. `vendor` can change value from one loop to the next, but only the last value of vendor is deleted from the counter handler?

Reply:
I assumed that I would need to remove the pending request from the queue in case the user terminates the job request before a license is allocated to it. But I should introduce a check inside the `licenseCounter.del()` method to check whether the vendor exists. The vendor will not change in between iterations because the server is receiving the same repeated request from the client when the client polls the server with a license request.

Reply:
Genuinely asking, is it ok to call `licenseCounter.del` with empty hash and empty vendor? It can happen from multiple threads, if we have an error on the first Recv.

Also, "the vendor will not change in between iterations" is entirely up to what the client is sending? Having server code that can end up in inconsistent state depending on the client behavior worries me. 

Impact here is low, so I'm ok with this PR, it's a first step. But writing the code so that it is resilient to this kind of problems is not significantly harder, I claim the code here would even be simpler if eg, only considered the vendor on first message, eliminated the hash, ...